### PR TITLE
[KYUUBI #6265] Resource isolation in Spark Scala mode

### LINKED ISSUE #6265: [TASK][MEDIUM] Resource isolation in Spark Scala mode
### What's the level of this task?

MEDIUM

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before creating

- [X] I have searched in the [task list](https://github.com/orgs/apache/projects/343) and found no similar tasks.

### Mentor

- [ ] I have sufficient expertise on this task, and I volunteer to be a mentor of this task to guide contributors through the task.

### Skill requirements

- Knowledge of Spark/Scala REPL
- Knowledge of classloader

### Background and Goals

The current Kyuubi Scala mode reuses the code of Spark Shell, while Spark Shell is designed for single session and client deploy mode, which has certain limitations for Kyuubis multi-tenant use cases.

Thanks to Spark Connect, a similar mechanism is required to support isolated sessions (i.e. isolated resources/classloader) so that each REPL user does not conflict with other users on the same cluster.

As of Spark 3.5, the core/sql already has infrastructure to support isolated resources/classloader, see details in

- SPARK-44078
- SPARK-44146
- SPARK-45856

### Implementation steps

Investigate how could Kyuubi Spark Scala mode leverage those infrastructures to support session-level isolated resources/classloader in Spark 3.5+

### Additional context

Introduction of [2024H1 Kyuubi Code Contribution Program](https://github.com/apache/kyuubi/issues/6232)

### PR DESCRIPTION
**Pull request by beryllw**
# :mag: Description
## Issue References 🔗

This pull request fixes #6265 

## Describe Your Solution 🔧

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Related Unit Tests

---

# Checklist 📝

- [ ] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

**Be nice. Be informative.**

### CODE REVIEW

**Comment by pan3793** [comment]
can you summarize your design/change on the PR description?

**Comment by beryllw** [comment]
> can you summarize your design/change on the PR description?

I've been quite busy recently,, sorry for not replying in time.

For now `ExecuteScala` use `spark.sharedState.jarClassLoader` to get the URLs for loading classes and resources, and then repl loaded.`SharedState` do not support isolated class loaded.

SparkSession use `HiveSessionResourceLoader` to support isolated class loaded.So we can get the URLs for loading classes and resources by `client.clientLoader.classLoader.getURLs`.

Spark 3.5 use `JobArtifactSet` to support add jar/resources to different path.

We can use it to support add resources isolated.And then repl loaded isolated classes/resources from `HiveSessionResourceLoader`.

**Comment by beryllw** [comment]
Only Spark 3.5 supports `JobArtifactState`. I need to use reflection to solve the compatibility problem, but I don’t know how to use reflection to call `def withActiveJobArtifactState[T](state: JobArtifactState)(block: => T): T = {}`.

**Comment by beryllw** [comment]
@pan3793 cc, and need some help, thank you.
